fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/JoinedRecordTypeTest.java

```diff
@@ -0,0 +1,208 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Tests of the {@link JoinedRecordType} class. Some of these tests may require access to an underlying FDB record
+ * store.
+ */
+@Tag(Tags.RequiresFDB)
+public class JoinedRecordTypeTest extends FDBRecordStoreQueryTestBase {
+
+    public static final String JOINED_RECORD_NAME = "JoinedRecord";
+    public static final String SIMPLE_RECORD = "simple_record";
+    public static final String OTHER_RECORD = "other_record";
+
+    @Nonnull
+    private static RecordMetaData baseMetaData(@Nonnull RecordMetaDataHook hook) {
+        RecordMetaDataBuilder metaDataBuilder = RecordMetaData.newBuilder()
+                .setRecords(TestRecordsJoinIndexProto.getDescriptor());
+        hook.apply(metaDataBuilder);
+        return metaDataBuilder.build();
+    }
+
+    @Nonnull
+    private static RecordMetaDataHook addJoinedType() {
+        return metaDataBuilder -> {
+            JoinedRecordTypeBuilder typeBuilder = metaDataBuilder.addJoinedRecordType(JOINED_RECORD_NAME);
+            typeBuilder.addConstituent(SIMPLE_RECORD, "MySimpleRecord");
+            typeBuilder.addConstituent(OTHER_RECORD, "MyOtherRecord");
+            typeBuilder.addJoin(SIMPLE_RECORD, field("other_rec_no"), OTHER_RECORD, field("rec_no"));
+        };
+    }
+
+    @Test
+    void loadSyntheticRecord() throws ExecutionException, InterruptedException {
+        final Tuple joinedPrimaryKey;
+        final FDBStoredRecord<Message> simpleRecord;
+        final FDBStoredRecord<Message> otherRecord;
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenRecordStore(context, baseMetaData(addJoinedType()));
+
+            Tuple syntheticRecordTypeKey = recordStore.getRecordMetaData()
+                    .getSyntheticRecordType(JOINED_RECORD_NAME)
+                    .getRecordTypeKeyTuple();
+            simpleRecord = recordStore.saveRecord(createSimpleRecord(100, 10));
+            otherRecord = recordStore.saveRecord(createOtherRecord(101, 11));
+            joinedPrimaryKey = Tuple.from(
+                    syntheticRecordTypeKey.getItems().get(0),
+                    simpleRecord.getPrimaryKey().getItems(),
+                    otherRecord.getPrimaryKey().getItems());
+
+            context.commit();
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            createOrOpenRecordStore(context, baseMetaData(addJoinedType()));
+
+            final FDBSyntheticRecord rec = recordStore.loadSyntheticRecord(joinedPrimaryKey).get();
+            assertEquals(2, rec.getConstituents().size());
+
+            assertEquals(simpleRecord, rec.getConstituent(SIMPLE_RECORD));
+            assertEquals(otherRecord, rec.getConstituent(OTHER_RECORD));
+
+            context.commit();
+        }
+    }
+
+    @Test
+    void loadSyntheticRecordFailsMissingConstituent() throws ExecutionException, InterruptedException {
+        final Tuple joinedPrimaryKey;
```

Comment:
All three of these tests have basically the same setup, it seems like it would be nice to extract this into a helper class (probably nested) that saves the two types.

Reply:
On the other hand, you could probably combine these two into a parameterized test with two booleans
`loadSyntheticRecordFailsMissingConstituent(boolean deleteSimple, boolean deleteOther)`
and then run with all the combinations, and make sure that if either is deleted you end up with the appropriate behaviors.

Reply:
Yes, done - one method for positive test and one for failing test